app/services/setting_registry.rb

```diff
@@ -0,0 +1,107 @@
+class SettingRegistry
+  include Singleton
+
+  def initialize
+    @settings = {}
+  end
+
+  def ready?
+    @settings.any?
+  end
+
+  def logger
+    Foreman::Logging.logger('app')
+  end
+
+  def find(name)
+    logger.warn("Setting is not initialized yet, requested value for #{name} will be always nil") unless ready?
+    @settings[name.to_s]
+  end
+
+  # Returns a setting effective value
+  # Call as `Foreman.settings[<name>]`
+  #
+  #   Foreman.settings['default_locale'] => nil.
+  #   Foreman.settings[:default_locale] => nil.
+  #
+  def [](name)
+    definition = find(name)
+    unless definition
+      logger.warn("Setting #{name} has no definition, please define it before using") if ready?
+      return
+    end
+    definition.value
+  end
+
+  def []=(name, value)
+    definition = find(name)
+    raise ::Foreman::Exception.new(N_("Setting definition for '%s' not found, can not set"), name) unless definition
+    db_record = _find_or_new_db_record(name)
+    db_record.value = value
+    db_record.save!
+    definition.value = db_record.value
+  end
+
+  def load
+    # add() all setting definitions
+    Setting.descendants.each do |cat_cls|
+      if cat_cls.default_settings.empty?
+        # Setting category uses really old way of doing things
+        _load_category_from_db(cat_cls)
+      else
+        cat_cls.default_settings.each { |s| _add(s[:name], s.except(:name).merge(category: cat_cls.name)) }
+      end
+    end
+
+    # load all db values
+    load_values
+
+    # if create_missing create missing values in DB
+    # not needed now as old load mechanism takes care of that
+  end
+
+  def load_values
+    loaded_names = []
+    Setting.unscoped.all.each do |set|
+      set_definition = find(set.name)
+      logger.debug("Setting #{set.name} has no definition, clean up your database") and next unless set_definition
+      loaded_names << set.name
+      set_definition.updated_at = set.updated_at
+      set_definition.value = set.value
```

Comment:
Shouldn't this check if the value changed and only then update it?
```suggestion
      next if set_definition.value == set.value

      set_definition.updated_at = set.updated_at
      set_definition.value = set.value
```

Reply:
I don't believe so, I see no value in such if, we did load the model already, there is no reason not setting the value even if it is still the same, it is simple setter on a class, so it is probably even faster just setting it directly.

Reply:
My reasoning is that you don't want to **always** set `updated_at`. Otherwise you may end up with useless values. Maybe that fear isn't really a problem.

Reply:
I don't understand :thinking:
`set` is a setting model, so we just set the `updated_at` of the presenter to a value we have in database, so if the value did not change in the meantime, the `updated_at` timestamp will be the same, we just pass the value, so we don't keep the whole ActiveRecord model in memory just for proxying the `value` and `updated_at` methods to it.

Reply:
Ah right, I was confused. I somehow assumed the definition was a model instance. It makes sense you don't want to keep the model reference.